### ISSUE TITLE
Bookmarks are uploaded to sync even if I opt out during account creation

### ISSUE THREAD

**Issue by rfk** [mozilla-mobile/firefox-ios#6348]
I encountered this behaviour on both current master, and the current tip of the `v24.x` branch (that is, with both the pre-rust-fxa and post-rust-fxa codebases).

## Steps to reproduce

* Sign in to sync, and opt to create a new account.
* When ask what datatypes you'd like to sync, uncheck "bookmarks".
* Complete the signin process.
* Go to the account settings screen.
  * Observe that "bookmarks" is checked, despite unchecking it during account signup.
* Watch log output as the browser performs its first sync.
  * Observe that sync appears to be uploading my bookmarks, despite me telling it not to.
* Exit the account settings screen.
* Click the "sync now" button.
* Go back into the account settings screen.
  * Observe that "bookmarks" is now unchecked after having done a second sync.

### Expected behavior

* The first sync should not have uploaded my bookmarks at all, because I opted out during account creation.

### Actual behavior

* The browser needlessly uploads my bookmarks to the sync server on its first sync, ~~then deletes them on its second sync.~~
    * Update: I was mistaken about them being deleted on the next sync, AFAICT they get left on the server.
* Hooray for encrypting everything by default 😁 

### Device & build information

I'm mucking around with latest master in the simulator.

**Comment by rfk** [comment]
I was using a test account so I figure there's no harm in me sharing the log output from this run. Here's the full output: [ios_bookmarks_sync_logs.txt](https://github.com/mozilla-mobile/firefox-ios/files/4412624/ios_bookmarks_sync_logs.txt)

And my thoughts as I look through what happened. The first sync begins with:

```
2020-04-01 14:05:35.102 [Info] [Profile.swift:1127] syncWith(synchronizers:statsSession:why:) > Syncing ["clients", "tabs", "bookmarks", "history", "logins"]
2020-04-01 14:05:35.103 [Debug] [Profile.swift:1134] syncWith(synchronizers:statsSession:why:) > engines to enable: []
2020-04-01 14:05:35.103 [Debug] [Profile.swift:1135] syncWith(synchronizers:statsSession:why:) > engines to disable: ["bookmarks"]
```

Which is off to a good start, indicating that it intends to disable the bookmarks engine.

```

2020-04-01 14:05:36.086 [Info] [SyncStateMachine.swift:188] toReady(_:) > Advancing to InitialWithLiveToken.
2020-04-01 14:05:36.086 [Info] [SyncStateMachine.swift:329] init(scratchpad:token:) > Inited initialWithLiveToken
2020-04-01 14:05:36.087 [Info] [SyncStateMachine.swift:122] advanceFromState(_:) > advanceFromState: InitialWithLiveToken
2020-04-01 14:05:36.351 [Debug] [StorageClient.swift:381] failFromResponse(_:) > Status code: 200.
2020-04-01 14:05:36.352 [Info] [SyncStateMachine.swift:329] init(scratchpad:token:) > Inited initialWithLiveTokenAndInfo
2020-04-01 14:05:36.352 [Info] [SyncStateMachine.swift:122] advanceFromState(_:) > advanceFromState: InitialWithLiveTokenAndInfo
2020-04-01 14:05:36.352 [Debug] [SyncStateMachine.swift:787] advance() > No cached meta/global found. Fetching fresh meta/global.
2020-04-01 14:05:36.352 [Info] [SyncStateMachine.swift:313] init(client:scratchpad:token:) > Inited needsFreshMetaGlobal
2020-04-01 14:05:36.352 [Info] [SyncStateMachine.swift:122] advanceFromState(_:) > advanceFromState: NeedsFreshMetaGlobal
2020-04-01 14:05:36.601 [Debug] [StorageClient.swift:381] failFromResponse(_:) > Status code: 404.
2020-04-01 14:05:36.602 [Debug] [StorageClient.swift:391] failFromResponse(_:) > NotFound
2020-04-01 14:05:36.602 [Info] [SyncStateMachine.swift:122] advanceFromState(_:) > advanceFromState: MissingMetaGlobal
2020-04-01 14:05:36.602 [Info] [SyncStateMachine.swift:122] advanceFromState(_:) > advanceFromState: FreshStartRequired
2020-04-01 14:05:36.807 [Debug] [StorageClient.swift:381] failFromResponse(_:) > Status code: 200.
2020-04-01 14:05:36.808 [Info] [SyncStateMachine.swift:122] advanceFromState(_:) > advanceFromState: ServerConfigurationRequired
2020-04-01 14:05:36.810 [Debug] [State.swift:583] pickle(_:) > Removing keys from Keychain.
2020-04-01 14:05:37.041 [Debug] [StorageClient.swift:381] failFromResponse(_:) > Status code: 200.
2020-04-01 14:05:37.280 [Debug] [StorageClient.swift:381] failFromResponse(_:) > Status code: 200.
```

This is a fresh account, so the sync quickly discovers that it needs to upload new server state. The code for the `ServerConfigurationRequired` state does indeed construct and upload a fresh metaglobal record, but it doesn't log enough info to tell me what the contents were. I'll try to do another run with more logging.

A whole bunch of additional sync setup logging happens, then we get to:

```

2020-04-01 14:05:38.648 [Debug] [Profile.swift:1146] syncWith(synchronizers:statsSession:why:) > Syncing bookmarks…
2020-04-01 14:05:38.649 [Debug] [Profile.swift:986] syncBookmarksWithDelegate(_:prefs:ready:why:) > Syncing bookmarks to storage.
2020-04-01 14:05:38.658 [Info] [Profile.swift:270] init(localName:syncDelegate:clear:) > [RUST][sync15::sync_multiple] Syncing 1 stores
2020-04-01 14:05:38.658 [Info] [Profile.swift:270] init(localName:syncDelegate:clear:) > [RUST][sync15::sync_multiple] Loading/initializing persisted state
2020-04-01 14:05:38.658 [Info] [Profile.swift:270] init(localName:syncDelegate:clear:) > [RUST][sync15::sync_multiple] Preparing client info
2020-04-01 14:05:38.658 [Info] [Profile.swift:270] init(localName:syncDelegate:clear:) > [RUST][sync15::sync_multiple] Entering sync state machine
2020-04-01 14:05:38.658 [Info] [Profile.swift:270] init(localName:syncDelegate:clear:) > [RUST][sync15::sync_multiple] Advancing state machine to ready (full)
2020-04-01 14:05:39.659 [Info] [Profile.swift:270] init(localName:syncDelegate:clear:) > [RUST][sync15::client] Successful request to "/1.5/141552812/info/configuration", incoming x-last-modified=ServerTimestamp(1585710338540)
2020-04-01 14:05:39.858 [Info] [Profile.swift:270] init(localName:syncDelegate:clear:) > [RUST][sync15::client] Successful request to "/1.5/141552812/info/collections", incoming x-last-modified=ServerTimestamp(1585710338540)
2020-04-01 14:05:40.056 [Info] [Profile.swift:270] init(localName:syncDelegate:clear:) > [RUST][sync15::client] Successful request to "/1.5/141552812/storage/meta/global", incoming x-last-modified=ServerTimestamp(1585710336930)
2020-04-01 14:05:40.056 [Info] [Profile.swift:270] init(localName:syncDelegate:clear:) > [RUST][sync15::state] Have info/collections and meta/global. Computing new engine states
2020-04-01 14:05:40.254 [Info] [Profile.swift:270] init(localName:syncDelegate:clear:) > [RUST][sync15::client] Successful request to "/1.5/141552812/storage/crypto/keys", incoming x-last-modified=ServerTimestamp(1585710337160)
2020-04-01 14:05:40.255 [Info] [Profile.swift:270] init(localName:syncDelegate:clear:) > [RUST][sync15::sync_multiple] Synchronizing stores
2020-04-01 14:05:40.255 [Info] [Profile.swift:270] init(localName:syncDelegate:clear:) > [RUST][sync15::sync_multiple] Syncing bookmarks engine!
2020-04-01 14:05:40.255 [Info] [Profile.swift:270] init(localName:syncDelegate:clear:) > [RUST][sync15::sync] Syncing collection bookmarks
2020-04-01 14:05:40.256 [Info] [Profile.swift:270] init(localName:syncDelegate:clear:) > [RUST][sync15::coll_state] Resetting bookmarks store
2020-04-01 14:05:40.259 [Info] [Profile.swift:270] init(localName:syncDelegate:clear:) > [RUST][sync15::sync] skipping incoming for bookmarks - not needed.
2020-04-01 14:05:40.262 [Info] [Profile.swift:270] init(localName:syncDelegate:clear:) > [RUST][sync15::sync] Uploading 13 outgoing changes
2020-04-01 14:05:40.263 [Info] [Profile.swift:270] init(localName:syncDelegate:clear:) > [RUST][sync15::request] Posting 13 records of 6412 bytes
2020-04-01 14:05:40.542 [Info] [Profile.swift:270] init(localName:syncDelegate:clear:) > [RUST][sync15::client] Successful request to "/1.5/141552812/storage/bookmarks", incoming x-last-modified=ServerTimestamp(1585710340380)
2020-04-01 14:05:40.542 [Info] [Profile.swift:270] init(localName:syncDelegate:clear:) > [RUST][sync15::sync] Upload success (13 records success, 0 records failed)
2020-04-01 14:05:40.547 [Info] [Profile.swift:270] init(localName:syncDelegate:clear:) > [RUST][sync15::sync] Sync finished!
2020-04-01 14:05:40.547 [Info] [Profile.swift:270] init(localName:syncDelegate:clear:) > [RUST][sync15::sync_multiple] Sync of bookmarks was successful!
2020-04-01 14:05:40.547 [Info] [Profile.swift:270] init(localName:syncDelegate:clear:) > [RUST][sync15::sync_multiple] Finished syncing stores.
```

This is the rust bookmarks engine, and it's definitely syncing my bookmarks. We see here that it has successfully fetched metaglobal and it believes that the contents indicate that bookmarks should be synced. Maybe there's something wrong with the logic here that's not correctly detecting that bookmarks should be disabled?

The other stores then sync as expected, as the overall sync succeeds and uploads some telemetry:

```
2020-04-01 14:05:41.833 [Debug] [SyncTelemetry.swift:52] send(ping:docType:) > Ping payload: {"os":{"name":"iOS","version":"13.4","locale":"en_US"},"version":1,"uid":"b14fb723a99dee89d460fa7618c026d6","syncs":[{"devices":[],"engines":[{"outgoing":{"failed":0,"sent":1},"took":455,"name":"clients"},{"name":"tabs","took":222,"outgoing":{"failed":0,"sent":1}},{"name":"bookmarks","took":0},{"outgoing":{"sent":19,"failed":0},"name":"history","took":262},{"took":0,"name":"logins"}],"when":1585710337968,"didLogin":true,"why":"didLogin","took":3853}],"deviceID":"16d3faf6b06659cb0203d0cb066aa602aa8ce24be179ed6b1541eee98322f9f5","why":"schedule","events":[]}
2020-04-01 14:05:42.670 [Debug] [SyncTelemetry.swift:82] send(ping:docType:) > Ping response: 200.
```

Interestingly, the telemetry here says `{"name":"bookmarks","took":0}`, but we definitely saw the bookmarks engine syncing, so perhaps the telemetry here is not measuring the right thing? The logins engine, which is also in rust, also claims to have taken no time to complete.

While I was poking at the output here, the browser's next sync kicked off in the background, and we see:

```

2020-04-01 14:14:20.097 [Info] [Profile.swift:1127] syncWith(synchronizers:statsSession:why:) > Syncing ["clients", "tabs", "bookmarks", "history", "logins"]
2020-04-01 14:14:20.097 [Info] [SyncAuthState.swift:84] token(_:canBeExpired:) > Returning cached token, which should be valid.
2020-04-01 14:14:20.098 [Debug] [SyncStateMachine.swift:148] toReady(_:) > Got token from auth state.
2020-04-01 14:14:20.101 [Debug] [State.swift:467] unpickleV1FromPrefs(_:syncKeyBundle:) > Read keys from Keychain with label -dEcadql8IPC.
2020-04-01 14:14:20.101 [Info] [SyncStateMachine.swift:188] toReady(_:) > Advancing to InitialWithLiveToken.
2020-04-01 14:14:20.101 [Info] [SyncStateMachine.swift:329] init(scratchpad:token:) > Inited initialWithLiveToken
2020-04-01 14:14:20.101 [Info] [SyncStateMachine.swift:122] advanceFromState(_:) > advanceFromState: InitialWithLiveToken
2020-04-01 14:14:20.353 [Debug] [StorageClient.swift:381] failFromResponse(_:) > Status code: 200.
2020-04-01 14:14:20.354 [Info] [SyncStateMachine.swift:329] init(scratchpad:token:) > Inited initialWithLiveTokenAndInfo
2020-04-01 14:14:20.354 [Info] [SyncStateMachine.swift:122] advanceFromState(_:) > advanceFromState: InitialWithLiveTokenAndInfo
2020-04-01 14:14:20.354 [Debug] [SyncStateMachine.swift:770] advance() > Cached meta/global fetched at 1585710337640, newer than server modified 1585710336930. Using cached meta/global.
2020-04-01 14:14:20.354 [Info] [SyncStateMachine.swift:313] init(client:scratchpad:token:) > Inited hasMetaGlobal
2020-04-01 14:14:20.354 [Info] [SyncStateMachine.swift:122] advanceFromState(_:) > advanceFromState: HasMetaGlobal
2020-04-01 14:14:20.624 [Debug] [StorageClient.swift:381] failFromResponse(_:) > Status code: 200.
2020-04-01 14:14:20.625 [Debug] [State.swift:578] pickle(_:) > Storing keys in Keychain with label keys.-dEcadql8IPC.
2020-04-01 14:14:20.629 [Info] [SyncStateMachine.swift:313] init(client:scratchpad:token:) > Inited needsFreshMetaGlobal
2020-04-01 14:14:20.630 [Info] [SyncStateMachine.swift:122] advanceFromState(_:) > advanceFromState: NeedsFreshMetaGlobal
2020-04-01 14:14:20.827 [Debug] [StorageClient.swift:381] failFromResponse(_:) > Status code: 200.
2020-04-01 14:14:20.828 [Info] [SyncStateMachine.swift:313] init(client:scratchpad:token:) > Inited resolveMetaGlobalVersion
2020-04-01 14:14:20.828 [Info] [SyncStateMachine.swift:122] advanceFromState(_:) > advanceFromState: ResolveMetaGlobalVersion
2020-04-01 14:14:20.828 [Info] [SyncStateMachine.swift:313] init(client:scratchpad:token:) > Inited resolveMetaGlobalContent
2020-04-01 14:14:20.828 [Info] [SyncStateMachine.swift:122] advanceFromState(_:) > advanceFromState: ResolveMetaGlobalContent
2020-04-01 14:14:20.829 [Info] [SyncStateMachine.swift:702] advance() > Remote meta/global disabled previously enabled engine bookmarks.
```

Now that's interesting, it believes that *some other client* has updated metaglobal to disable syncing of bookmarks. But there aren't any other clients, so something has correctly uploaded a metaglobal with bookmarks in declined.

When we reach the rust bookmarks sync engine, it has this to say:

```

2020-04-01 14:14:21.305 [Debug] [Profile.swift:1146] syncWith(synchronizers:statsSession:why:) > Syncing bookmarks…
2020-04-01 14:14:21.305 [Debug] [Profile.swift:986] syncBookmarksWithDelegate(_:prefs:ready:why:) > Syncing bookmarks to storage.
2020-04-01 14:14:21.873 [Info] [Profile.swift:270] init(localName:syncDelegate:clear:) > [RUST][sync15::sync_multiple] Syncing 1 stores
2020-04-01 14:14:21.873 [Info] [Profile.swift:270] init(localName:syncDelegate:clear:) > [RUST][sync15::sync_multiple] Loading/initializing persisted state
2020-04-01 14:14:21.873 [Info] [Profile.swift:270] init(localName:syncDelegate:clear:) > [RUST][sync15::sync_multiple] Preparing client info
2020-04-01 14:14:21.873 [Info] [Profile.swift:270] init(localName:syncDelegate:clear:) > [RUST][sync15::sync_multiple] Entering sync state machine
2020-04-01 14:14:21.873 [Info] [Profile.swift:270] init(localName:syncDelegate:clear:) > [RUST][sync15::sync_multiple] Advancing state machine to ready (full)
2020-04-01 14:14:22.150 [Info] [Profile.swift:270] init(localName:syncDelegate:clear:) > [RUST][sync15::client] Successful request to "/1.5/141552812/info/collections", incoming x-last-modified=ServerTimestamp(1585710860520)
2020-04-01 14:14:22.355 [Info] [Profile.swift:270] init(localName:syncDelegate:clear:) > [RUST][sync15::client] Successful request to "/1.5/141552812/info/collections", incoming x-last-modified=ServerTimestamp(1585710860520)
2020-04-01 14:14:22.550 [Info] [Profile.swift:270] init(localName:syncDelegate:clear:) > [RUST][sync15::client] Successful request to "/1.5/141552812/storage/meta/global", incoming x-last-modified=ServerTimestamp(1585710860520)
2020-04-01 14:14:22.550 [Info] [Profile.swift:270] init(localName:syncDelegate:clear:) > [RUST][sync15::state] Have info/collections and meta/global. Computing new engine states
2020-04-01 14:14:22.550 [Warning] [Profile.swift:272] init(localName:syncDelegate:clear:) > [RUST][sync15::state] Remote state contains engines which are in both info/collections and meta/global's declined: {"bookmarks"}
2020-04-01 14:14:22.754 [Info] [Profile.swift:270] init(localName:syncDelegate:clear:) > [RUST][sync15::client] Successful request to "/1.5/141552812/storage/crypto/keys", incoming x-last-modified=ServerTimestamp(1585710337160)
2020-04-01 14:14:22.754 [Info] [Profile.swift:270] init(localName:syncDelegate:clear:) > [RUST][sync15::sync_multiple] Resetting engine bookmarks, as it was declined remotely
2020-04-01 14:14:22.756 [Info] [Profile.swift:270] init(localName:syncDelegate:clear:) > [RUST][sync15::sync_multiple] Synchronizing stores
2020-04-01 14:14:22.756 [Info] [Profile.swift:270] init(localName:syncDelegate:clear:) > [RUST][sync15::sync_multiple] The bookmarks engine is declined. Skipping
2020-04-01 14:14:22.756 [Info] [Profile.swift:270] init(localName:syncDelegate:clear:) > [RUST][sync15::sync_multiple] Finished syncing stores.
```

So it also correctly identifies that the bookmarks engine has been disabled. It also logs a warning that the bookmarks collection still exists, but AFAICT it doesn't make any attempt to clean up the data.

The rest of the sync seems to complete normally, and if I go into my account settings I can see that bookmarks are no longer being synced.

**Comment by rfk** [comment]
> The browser needlessly uploads my bookmarks to the sync server on its first sync, then deletes them on its second sync.

I was wrong about this, it leaves my encrypted bookmark records on the server.

**Comment by rfk** [comment]
> Now that's interesting, it believes that some other client has updated metaglobal to disable syncing
> of bookmarks. But there aren't any other clients, so something has correctly uploaded a metaglobal
> with bookmarks in declined.

I added some extra logging to understand what's happening here.  It turns out that on its second sync, the device notices that it has some pending engine-enablement changes, so it uploads a new (correct this time) record as part of advancing from the `HasMetaGlobal` state.

**Comment by SimonBasca** [comment]
Verified as fixed on beta 25.0 (17758) and master 8c79326a3e40a.
Tested the following scenarios: 

1. Create account 
2. When ask what datatypes you'd like to sync, uncheck "bookmarks"
3. Check the account settings screen. Bookmarks will be unchecked.

Also tested this scenario for History, Tabs and Logins.
